umbral/umbral.py

```diff
@@ -115,98 +116,135 @@ def __bytes__(self):
         h = hash_to_bn([e, v], params)
 
         return params.g * s == v + (e * h)
 
     def attach_cfrag(self, cfrag: CapsuleFrag):
         self.cfrags[cfrag.bn_kfrag_id] = cfrag
 
-    def reconstruct(self):
+    def original_ephemeral_points(self):
+        return self._point_eph_e + self._point_eph_v
+
+    def reconstructed_ephemeral_points(self):
+        return self._point_eph_e_prime + self._point_eph_v_prime
+
+    def open(self, pub_bob, priv_bob, pub_alice, force_reopen=False, pre=None):
+        # TODO: Raise an error here if Bob has gathered enough cFrags.
+        if self.contents and not force_reopen:
+            newly_opened = True
+        else:
+            self._reconstruct(pre=pre)
+            if not pre:
+                pre = PRE(UmbralParameters())
+            self._contents = pre._decapsulate_reencrypted(pub_bob, priv_bob, pub_alice, self)
+            newly_opened = False
+        return self.contents, newly_opened
+
+    def original_components(self):
+        return self._point_eph_e, self._point_eph_v, self._bn_sig
+
+    def _reconstruct(self, pre):
         id_cfrag_pairs = list(self.cfrags.items())
         id_0, cfrag_0 = id_cfrag_pairs[0]
         if len(id_cfrag_pairs) > 1:
-            ids = self.cfrags.keys() 
+            ids = self.cfrags.keys()
             lambda_0 = lambda_coeff(id_0, ids)
             e = cfrag_0.point_eph_e1 * lambda_0
             v = cfrag_0.point_eph_v1 * lambda_0
-            
-            for id_i,cfrag in id_cfrag_pairs[1:]:
+
+            for id_i, cfrag in id_cfrag_pairs[1:]:
                 lambda_i = lambda_coeff(id_i, ids)
                 e = e + (cfrag.point_eph_e1 * lambda_i)
                 v = v + (cfrag.point_eph_v1 * lambda_i)
         else:
             e = cfrag_0.point_eph_e1
             v = cfrag_0.point_eph_v1
-        
-        return ReconstructedCapsule(e_prime=e, v_prime=v, x=cfrag_0.point_eph_ni)
-
-    def __bytes__(self):
-        self.to_bytes()
-
 
-class ReconstructedCapsule(object):
-    def __init__(self, e_prime, v_prime, x):
-        self.point_eph_e_prime = e_prime
-        self.point_eph_v_prime = v_prime
-        self.point_eph_ni = x
+        self._point_eph_e_prime = e
+        self._point_eph_v_prime = v
+        self._point_noninteractive = cfrag_0.point_eph_ni
 
-    @staticmethod
-    def from_bytes(data: bytes, curve: ec.EllipticCurve):
+    def _reconstructed_bytes(self):
         """
-        Instantiate ReconstructedCapsule from serialized data.
+        Serialize the reconstruction components into a bytestring.
         """
-        e_prime = Point.from_bytes(data[0:33], curve)
-        v_prime = Point.from_bytes(data[33:66], curve)
-        eph_ni = Point.from_bytes(data[66:99], curve)
+        eph_e = self._point_eph_e_prime
+        eph_v = self._point_eph_v_prime
+        point_noninter = self._point_noninteractive
 
-        return ReconstructedCapsule(e_prime, v_prime, eph_ni)
-
-    def to_bytes(self):
```

Comment:
Why are you getting rid of `to_bytes`? Keep this for uniformity, it's how cryptographic APIs handle their bytestring representations.